### PR TITLE
Add custom unmarshaling for ForwardMessage

### PR DESCRIPTION
**Pull request by ericmillin**
This PR:
- Adds a `MessageOptions` struct
- Changes the Records type to `map[string]interface{}` to allow us to test using captured data from a live FB instance
- Moves `ForwardMessage` to a new file
- Adds custom decode/unmarshal logic to `ForwardMessage` to allow for optional Options val

### CODE REVIEW

File: fluent/protocol/transport.go [hunk 1]
```diff
@@ -193,21 +126,72 @@ func (e EntryList) Equal(e2 EntryList) bool {
...
+// of the entry is responsible for converting to UTC.
+//msgp:tuple Entry
+type Entry struct {
+	// Timestamp can contain the timestamp in either seconds or nanoseconds
+	Timestamp int64
+	// Record is the actual event record.
+	Record Record
+}
+
+type MessageOptions struct {
```
Comment (ericmillin) [inline fluent/protocol/transport.go:140]: This is new

File: fluent/protocol/forward_message_test.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/tinylib/msgp/msgp"
+
+	. "github.ibm.com/Observability/fluent-forward-go/fluent/protocol"
+)
+
+var _ = Describe("ForwardMessage", func() {
```
Comment (ericmillin) [inline fluent/protocol/forward_message_test.go:15]: This file is new

File: fluent/protocol/forward_message.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package protocol
+
+import "github.com/tinylib/msgp/msgp"
+
+// ForwardMessage is used in Forward mode to send multiple events in a single
+// msgpack array within a single request.
+//msgp:tuple ForwardMessage
+//msgp:decode ignore ForwardMessage
+//msgp:unmarshal ignore ForwardMessage
+type ForwardMessage struct {
```
Comment (ericmillin) [inline fluent/protocol/forward_message.go:10]: This file is new

**Comment by Christopher-Wolf-ibm** [comment]
Looks good to me.

File: fluent/protocol/forward_message_test.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+						Record: map[string]interface{}{
+							"foo":    "kablooie",
+							"george": "frank",
+						},
+					},
+				},
+			}
+		})
+
+		When("there are options", func() {
```
Comment (ericmillin) [inline fluent/protocol/forward_message_test.go:47]: need to change this title. Can probably get rid of the `When` altogether

File: fluent/protocol/forward_message.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	return nil
+}
+
+func (fm *ForwardMessage) UnmarshalMsg(bits []byte) ([]byte, error) {
+	var (
+		sz  uint32
+		err error
+	)
+
+	sz, bits, err = msgp.ReadArrayHeaderBytes(bits)
```
Comment (ericmillin) [inline fluent/protocol/forward_message.go:52]: can be combined with next line

Comment (hqzhangibm) [reply fluent/protocol/forward_message.go:52]: looks good

File: fluent/protocol/forward_message_test.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+				msg.Options = opts
+				b, err := msg.MarshalMsg(nil)
+				Expect(err).NotTo(HaveOccurred())
+
+				var unmfwd ForwardMessage
+				_, err = unmfwd.UnmarshalMsg(b)
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(msg.Options).To(Equal(opts))
+				Expect(msg.Entries[0].Record).To(HaveKeyWithValue("foo", "bar"))
```
Comment (ericmillin) [inline fluent/protocol/forward_message_test.go:58]: This should be testing `unmfwd` not msg.

**Comment by hqzhangibm** [comment]
everything looks good.

**Review by ScarletTanager — CHANGES_REQUESTED** [review]
See my comments.

File: fluent/protocol/transport.go [hunk 7]
```diff
@@ -216,7 +200,7 @@ func eventStream(entries EntryList) []byte {
 	var buf bytes.Buffer
 	w := msgp.NewWriter(&buf)
 	for _, e := range entries {
-		e.EncodeMsg(w)
+		_ = e.EncodeMsg(w)
```
[the remark this answers is not in the document]
Comment (ericmillin) [reply fluent/protocol/transport.go:203]: must... appease... Linterthulu... aaaaaahhhhhhhh...

File: fluent/protocol/transport.go [hunk 8]
```diff
@@ -233,21 +217,21 @@ func eventStream(entries EntryList) []byte {
...
 
 // TODO: This is not working correctly yet
 func NewCompressedPackedForwardMessage(
-	tag string, entries []EntryExt, opts MessageOptions,
+	tag string, entries []EntryExt, opts *MessageOptions,
 ) *CompressedPackedForwardMessage {
 	var buf bytes.Buffer
 	zw := gzip.NewWriter(&buf)
-	zw.Write(eventStream(entries))
+	_, _ = zw.Write(eventStream(entries))
```
[the remark this answers is not in the document]
Comment (ericmillin) [reply fluent/protocol/transport.go:229]: Will add a `TODO` as it involves changing the function signature.

**Comment by ericmillin** [comment]
@ScarletTanager Added comments based on review feedback

File: fluent/protocol/forward_message_test.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+			Expect(unmfwd.Entries[0].Record).To(HaveKeyWithValue("foo", "bar"))
+			Expect(unmfwd.Entries[0].Record).To(HaveKeyWithValue("george", "jungle"))
+			Expect(unmfwd.Entries[1].Timestamp.Time.Equal(msg.Entries[1].Timestamp.Time)).To(BeTrue())
+			Expect(unmfwd.Entries[1].Record).To(HaveKeyWithValue("foo", "kablooie"))
+			Expect(unmfwd.Entries[1].Record).To(HaveKeyWithValue("george", "frank"))
+		}
+
+		It("Marshals and unmarshals correctly", func() {
+			// Uncomment after adding custom encoding
+			// testMarshalling(fwdmsg, nil)
```
Comment (ScarletTanager) [inline fluent/protocol/forward_message_test.go:72]: @ericmillin - did we need to uncomment and execute this test, too?  your comment implies that we do.

File: fluent/protocol/forward_message_test.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+			Expect(unmfwd.Entries[0].Record).To(HaveKeyWithValue("foo", "bar"))
+			Expect(unmfwd.Entries[0].Record).To(HaveKeyWithValue("george", "jungle"))
+			Expect(unmfwd.Entries[1].Timestamp.Time.Equal(msg.Entries[1].Timestamp.Time)).To(BeTrue())
+			Expect(unmfwd.Entries[1].Record).To(HaveKeyWithValue("foo", "kablooie"))
+			Expect(unmfwd.Entries[1].Record).To(HaveKeyWithValue("george", "frank"))
+		}
+
+		It("Marshals and unmarshals correctly", func() {
+			// Uncomment after adding custom encoding
+			// testEncodingDecoding(fwdmsg, nil)
```
Comment (ScarletTanager) [inline fluent/protocol/forward_message_test.go:106]: As for line 72.

**Review by ScarletTanager — CHANGES_REQUESTED** [review]
Take a look at my questions, see if anything needs to be done with them.

File: fluent/protocol/forward_message_test.go (hunk 9 above)
Comment (ericmillin) [reply fluent/protocol/forward_message_test.go:72]: Custom encoding is a TODO. I can make that explicit in the comment or remove it.

Comment (ScarletTanager) [reply fluent/protocol/forward_message_test.go:72]: As long as the commenting is intentional, I'm fine.  Just wanted to check.

**Review by ScarletTanager — APPROVED** [review]
We're good now.

**Pull request merged**